web/pandas/pdeps/0014-string-dtype.md

```diff
@@ -0,0 +1,388 @@
+keep using the existing `pd.StringDtype()` for now to keep the changes as
+minimal as possible, leaving a larger overhaul of the dtype system (potentially
+including different constructor functions or namespace) for a future discussion.
+See [GH-58613](https://github.com/pandas-dev/pandas/issues/58613) for the full
+discussion.
+
+## Backward compatibility
+
+The most visible backwards incompatible change will be that columns with string
+data will no longer have an `object` dtype. Therefore, code that assumes
+`object` dtype (such as `ser.dtype == object`) will need to be updated. This
+change is done as a hard break in a major release, as warning in advance for the
+changed inference is deemed too noisy.
+
+To allow testing code in advance, the
+`pd.options.future.infer_string = True` option is available for users.
+
+Otherwise, the actual string-specific functionality (such as the `.str` accessor
+methods) should generally all keep working as is. By preserving the current
+missing value semantics, this proposal is also backwards compatible on this
+aspect.
+
+### For existing users of `StringDtype`
+
+Users of the existing `StringDtype` will see more backwards incompatible
+changes, though. In pandas 3.0, calling `pd.StringDtype()` (or specifying
+`dtype="string"`) will start returning the new default string dtype using `NaN`,
+while up to now this returned the string dtype using `pd.NA` introduced in
+pandas 1.0.
+
+For example, this code snippet returned the NA-variant of `StringDtype` with
+pandas 1.x and 2.x:
+
+```python
+>>> pd.Series(["a", "b", None], dtype="string")
+0      a
+1      b
+2   <NA>
+dtype: string
+```
+
+but will start returning the new default NaN-variant of `StringDtype` with
+pandas 3.0. This means that the missing value sentinel will change from `pd.NA`
+to `NaN`, and that operations will no longer return nullable dtypes but default
+numpy dtypes (see the "Missing value semantics" section above).
+
+While this change will be transparent in many cases (e.g. checking for missing
+values with `isna()`/`dropna()`/`fillna()` or filtering rows with the result of
+a string predicate method keeps working regardless of the sentinel), this can be
+a breaking change if users relied on the exact sentinel or resulting dtype. Since
+pandas 1.0, the string dtype has been promoted quite a bit, and so we expect
+that many users already have started using this dtype, even though officially
+still labeled as "experimental".
+
+To smooth the upgrade experience for those users, it is proposed to add a
+deprecation warning before 3.0 when such dtype is created, giving them two
+options:
+
+- If the user just wants to have a dedicated "string" dtype (or the better
+  performance when using pyarrow) but is fine with using the default NaN
+  semantics, they can add `pd.options.future.infer_string = True` to their code
+  to suppress the warning and already opt-in to the future behaviour of pandas
+  3.0.
```

Comment:
It doesn't appear to me this is sufficient. Users may already be opting into Pyarrow-backed strings using this option, right? In which case they will never see a warning that the NA value will change under this proposal.

We could add an option like `default_na_value`. It would only be utilized in cases where the NA value is unspecified. Users could specify `np.nan`, `pd.NA`, or leave it unspecified. When unspecified, we'd have two cases:

 - User not using `dtype="string"` et al. They would not see any warnings, and in 3.0 get NaN semantics (either Pyarrow or NumPy storage, depending on if Pyarrow is installed).
 - User using `dtype="string"` et al. They would see a warning that the default will change in 3.0 to NaN, and they can silence it by specifying what NA value they want.

If/when we are ready to switch over to NA semantics, this option could be used for that purpose as well.

I would also be okay with @WillAyd's proposal to use `str` for the new default, and keep `string` to use `pd.NA`, but prefer changing `string` to use NaN. While it is asking users already using NA-semantics to make a change, it seems to me the change will be straightforward.

Reply:
> Users may already be opting into Pyarrow-backed strings using this option, right? In which case they will never see a warning that the NA value will change under this proposal.

Which option do you mean exactly? The `pd.options.future.infer_string = True` one? In that case, that is fine and the goal is indeed they don't see a warning, because when enabling this option, you _already_ get a NaN-variant of the string dtype.

(remember that we are in the somewhat strange situation where I am writing a PDEP to formalize behaviour that is partly (the `pd.StringDtype(storage="pyarrow_numpy")`) already in a released pandas version .. but also partly not. So some things I describe in this PDEP are already in main (and would be removed if this PDEP was rejected, I assume), some other things in this PDEP describe new behaviours to implement)

> * User using `dtype="string"` et al. They would see a warning that the default will change in 3.0 to NaN, and they can silence it by specifying what NA value they want.

With the current proposal, they can already silence it by setting the `pd.options.future.infer_string = True`, _in case they want the NaN variant_. 
It's true that it might be nice that there is also an option for the other case (so they don't have to change every dtype specification in their code). But I think I strongly prefer that for the default case (the NaN variant), this user API for this does not involve explicitly choosing NaN (like `pd.options.mode.default_na_value = np.nan`), because that is not a choice most users should make explicitly, it's simply the default missing value sentinel.

At some point we had an option to globally opt-in to our nullable NA-dtypes (https://github.com/pandas-dev/pandas/pull/50748, https://github.com/pandas-dev/pandas/issues/29752), but that was later removed again in favor of just keeping a keyword in the IO methods (https://github.com/pandas-dev/pandas/issues/51433, https://github.com/pandas-dev/pandas/pull/51853/).  
I think we will need such an option again when we start discussing how to migrate to all NA dtypes, but I would leave that until that point (if we did add an option right now that only worked for strings, then why wouldn't it already also work for the numeric and boolean nullable dtypes we also already have? That would seem strange to me)

Reply:
> I think we will need such an option again when we start discussing how to migrate to all NA dtypes, but I would leave that until that point

Yea I agree on keeping that out of scope. FWIW the latest revision of PDEP-13 proposes an `na_marker` option that defaults to pd.NA but can be set to "legacy" for np.nan / pd.NaT (and probably still pd.NA), so something we can discuss further there too. I was unaware of the historical option

Reply:
> Which option do you mean exactly? The `pd.options.future.infer_string = True` one? In that case, that is fine and the goal is indeed they don't see a warning, because when enabling this option, you _already_ get a NaN-variant of the string dtype.

:facepalm: I forgot it was the NaN variant. Thanks.

> At some point we had an option to globally opt-in to our nullable NA-dtypes... but that was later removed... I think we will need such an option again when we start discussing how to migrate to all NA dtypes, but I would leave that until that point

Agreed.